### PR TITLE
Feature/cls2 998 create job for eyb lead company contact matching

### PR DESCRIPTION
**Pull request by samuele-mattiuzzo**
### Checklist

* [ ] Has this branch been rebased on top of the current `main` branch?

  <details>
  <summary>Explanation</summary>
  
  The branch should not be stale or have conflicts at the time reviews are requested.
  
  </details>

* [ ] Is the CircleCI build passing?

### General points

<details>
<summary><strong>Other things to check</strong></summary><p></p>

* Make sure `fixtures/test_data.yaml` is maintained when updating models
* Consider the admin site when making changes to models
* Use select-/prefetch-related field lists in views and search apps, and update them when fields are added
* Make sure the README is updated e.g. when adding new environment variables

</details>

See [docs/CONTRIBUTING.md](https://github.com/uktrade/data-hub-api/blob/main/docs/CONTRIBUTING.md) for more guidelines.

### CODE REVIEW

File: datahub/investment_lead/services.py [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
 
 logger = logging.getLogger(__name__)
 
 
+def link_leads_to_companies():
+    queryset = get_leads_to_process()
+
+    for eyb_lead in queryset:
+        company = match_or_create_company_for_eyb_lead(eyb_lead)
+        if company:
```
Comment (samuele-mattiuzzo) [inline datahub/investment_lead/services.py:17]: this line does not need testing, the method below already handles the case of `company == None`; the if simply prevents an exception from being raised;

there's also to say that this PR is still a draft due to the fact that we haven't discussed how to handle unhappy paths yet.

File: datahub/investment_lead/test/test_services.py [hunk 2]
```diff
@@ -168,3 +170,47 @@ def test_create_contact_on_company(self):
         assert eyb_lead.company.contacts.count() == count + 1
         contact = eyb_lead.company.contacts.first()
         assert_eyb_lead_matches_contact(contact, eyb_lead)
+
+    def test_get_leads_to_process(self):
```
Comment (oliverjwroberts) [inline datahub/investment_lead/test/test_services.py:174]: Would be worth adding some more leads to test all the filters? For example:

```python
EYBLeadFactory(archived=True)
EYBLeadFactory(
    triage_hashed_uuid='a hashed uuid',
    user_hashed_uuid='another hashed uuid',
)
EYBLeadFactory(company=company)  # the company argument is not actually needed here because the factory calls the company factory by default
matching_hashed_uuid = generate_hashed_uuid()  # from investment_lead/test/factories.py
EYBLeadFactory(
    triage_hashed_uuid=matching_hashed_uuid,
    user_hashed_uuid=matching_hashed_uuid,
    company=None,
)
```

File: datahub/investment_lead/test/test_tasks/test_lead_linking_company_contact.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        triage_file_contents = file_contents_faker(records=triage_records)
+        user_file_contents = file_contents_faker(records=user_records)
+
+        setup_s3_bucket(
+            BUCKET,
+            [test_triage_file_path, test_user_file_path],
+            [triage_file_contents, user_file_contents],
+        )
+
+        with caplog.at_level(logging.INFO):
```
Comment (oliverjwroberts) [inline datahub/investment_lead/test/test_tasks/test_lead_linking_company_contact.py:84]: Minor, but you probably don't need these `with` statements if you aren't asserting any log messages in the `caplog` fixture.

File: datahub/investment_lead/test/test_tasks/test_lead_linking_company_contact.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        eyb_lead = EYBLead.objects.all()[0]
+        company = Company.objects.all()[0]
+        assert_eyb_lead_matches_company(company, eyb_lead)
+
+        assert eyb_lead.company.contacts.count() == 1
+        contact = eyb_lead.company.contacts.first()
+        assert_eyb_lead_matches_contact(contact, eyb_lead)
+
+    @mock_aws
+    def test_linking_existing_company_contact_success(
```
Comment (oliverjwroberts) [inline datahub/investment_lead/test/test_tasks/test_lead_linking_company_contact.py:103]: I wonder if for these tests, its worth directly testing the `link_leads_to_companies()` and `create_or_skip_eyb_lead_as_company_contact()` functions? Rather than setting up the bucket with files and calling the ingest triage and user functions (which are already tested). Then, in a separate test, we can check the `ingest_eyb_user_data` task calls the linking functions?

File: datahub/investment_lead/services.py (hunk 1 above)
Comment (oliverjwroberts) [reply datahub/investment_lead/services.py:17]: For now, shall we say if a lead doesn't have a company when we are expecting it to, we'll throw an error, skip the record, and log the exception? For example:

```python
def link_leads_to_companies():
    queryset = get_leads_to_process()
    for eyb_lead in queryset:
        try:
            match_or_create_company_for_eyb_lead(eyb_lead)
            create_or_skip_eyb_lead_as_company_contact(eyb_lead)
        except Exception as e:
            logger.error(f'Error linking EYB lead to company and contact: {e}. Skipping lead...')
            continue
```

**Review by dredmonds — APPROVED** [review]

File: datahub/investment_lead/services.py [hunk 5]
```diff
@@ -41,11 +66,11 @@ def add_new_company_from_eyb_lead(eyb_lead: EYBLead):
...
     company.duns_number = eyb_lead.duns_number
-    company.name = eyb_lead.company_name
-    company.sector = eyb_lead.sector
-    company.address_1 = eyb_lead.address_1
+    company.name = eyb_lead.company_name  # mandatory
+    company.sector = eyb_lead.sector  # mandatory
+    company.address_1 = eyb_lead.address_1  # mandatory
     company.address_2 = eyb_lead.address_2
-    company.address_town = eyb_lead.address_town
+    company.address_town = eyb_lead.address_town  # mandatory
```
Comment (marijnkampf) [inline datahub/investment_lead/services.py:73]: How are we/who is going to maintain these` # mandatory` comments when this changes on the model?

File: datahub/investment_lead/services.py [hunk 6]
```diff
@@ -1,12 +1,37 @@
...
 
+def link_leads_to_companies():
+    queryset = get_leads_to_process()
+
+    for eyb_lead in queryset:
+        try:
+            match_or_create_company_for_eyb_lead(eyb_lead)
+            create_or_skip_eyb_lead_as_company_contact(eyb_lead)
+        except Exception as e:
+            logger.error(f'Error linking EYB lead {eyb_lead.pk} to company/contact: {e}.')
```
Comment (samuele-mattiuzzo) [inline datahub/investment_lead/services.py:20]: decided for this error to be generic so that it's easier to search for it in the logs in case of need

**Review by marijnkampf — APPROVED** [review]

File: datahub/investment_lead/services.py (hunk 5 above)
Comment (samuele-mattiuzzo) [reply datahub/investment_lead/services.py:73]: Fair. I'll remove
